Add batch merge and refspec support to git task

ad22 opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

Add refspec support to git task corresponding to https://github.com/tektoncd/pipeline/pull/2320

Add a new batch merge task (https://github.com/tektoncd/pipeline/issues/508) that re-uses
the git-init container and subsequently performs the requested batch merge mode on top of
the checked out revision

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [X] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [X] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [?] Yaml file complies with [yamllint](https://github.com/adrienverge/yamllint) rules.

sbwsg commented:
Sorry for the extra work but would it be possible to split this into two PRs?  One for the refspec flag and another for the batch merge task?

The main reason for this is that the `refspec` flag won't be available from the `git-init` binary until your PR (https://github.com/tektoncd/pipeline/pull/2320) is merged and a new Pipelines release is cut.  If we merged this as-is then anyone `kubectl apply`ing the `git-clone` Task would end up with breakages because the `git-init` binary would complain about the missing flag: `flag provided but not defined: -refspec`. 

So ideally I'd like to be able to review and merge the batchmerge PR in isolation.

ad22 commented:
@sbwsg The batch merge functionality depends on the refspec feature being available, which is why I created them together. I can see the value of opening another PR for the batch merge, regardless, just because it addresses a separate requested feature.

Both features can wait till https://github.com/tektoncd/pipeline/pull/2320 is merged and released. Feel free to add the appropriate labels on this PR.

sbwsg commented:
/ok-to-test

ad22 commented:
Added tests, additional git merge/cherry-pick options and switched docker image to git-init:v0.12.0. Still waiting on the corresponding pipeline release, do not merge.

sbwsg commented on git/README.md:
````diff
@@ -135,3 +136,33 @@ spec:
       claimName: workspace-pvc
 ```
+
+## `git-batch-merge`
````
One minor suggestion here: it might be a good idea to explain the intended usage of this Task.  What's an example use case where this is useful?

ad22 replied:
Added a description

sbwsg replied:
Excellent, thank you!

sbwsg commented on git/git-batch-merge.yaml:
```diff
@@ -0,0 +1,117 @@
+      type: string
+    - name: revision
+      description: base git revision to checkout (branch, tag, sha, ref…)
+      default: master
```
suggest adding `type: string` in all the params where this is missing

ad22 replied:
Done

sbwsg commented:
@ad22 I've removed the hold label since 0.12 has been released.  Now's a good time to rebase on top of current `v1beta1` branch and address feedback if you'd still like to pursue this PR.  Thanks for your patience!

ad22 commented:
@sbwsg I've updated the PR. Its really a small wrapper around git-init. Please review the updated description, tektoncd/pipeline#508 and accept this if the you find it relevant. Thanks!

sbwsg commented on git/README.md:
```diff
@@ -135,3 +136,44 @@ spec:
+  (_default_: GitBatch.Task@tekton.dev). See [git-user-config][git-user-config]
+
+There is 1 additional result in addition to the one(s) mentioned for the git-clone task
+* **tree**: The [git tree][git-tree] object SHA1 that was created after batch merging the refs on HEAD.
```
Suggest just listing all the results this task emits, rather than referring to the results of another task.  This is so the user doesn't have to go looking for the results of another task to figure out what this one emits.  And there might come a time when either a) the git-clone task emits more results that this one does or b) we split these tasks into separate directories / readmes.

ad22 commented:
/test pull-tekton-catalog-build-tests

sbwsg commented:
/lgtm

sbwsg commented:
/test pull-tekton-catalog-build-tests
